service/src/main/kotlin/io/provenance/explorer/service/async/AsyncCaching.kt

```diff
@@ -177,6 +180,19 @@ class AsyncCaching(
         return TxUpdatedItems(addrs, markers)
     }
 
+    private fun saveEvents(txId: EntityID<Int>, tx: ServiceOuterClass.GetTxResponse, msg: com.google.protobuf.Any, msgId: Int, msgType: String, idx: Int) = transaction {
+        tx.txResponse.logsList.forEach { log ->
+            if (log.msgIndex == idx) {
```

Comment:
@crussell-figure Is this what you were thinking for making sure we only input the event once with the correct message rather than for every message... 🤦

Reply:
More like `tx.txResponse.logsList[idx]`. That way you only pull the logs that match the msg index. Then you can follow it with your forEarch, like `tx.txResponse.logsList[idx].eventsList.forEach`.